holmes/core/tool_calling_llm.py

```diff
@@ -568,31 +591,37 @@ def stream_analysis(it, peek_chunk):
 
                 for future in concurrent.futures.as_completed(futures):
                     tool_call_result: ToolCallResult = future.result()
-                    tool_response = tool_call_result.result.data
-                    if tool_call_result.result.status == ToolResultStatus.ERROR:
-                        tool_response = f"{tool_call_result.result.error or 'Tool execution failed'}:\n\n{tool_call_result.result.data or ''}".strip()
 
-                    if isinstance(tool_response, dict):
-                        tool_response = json.dumps(tool_response)
-                    if isinstance(tool_response, list):
-                        tool_response = json.dumps(tool_response)
+                    tool_response_content = (
+                        self._get_formatted_tool_call_response_content(tool_call_result)
+                    )
 
                     message_to_append = {
                         "tool_call_id": tool_call_result.tool_call_id,
                         "role": "tool",
                         "name": tool_call_result.tool_name,
-                        "content": tool_response,
+                        "content": tool_response_content,
                     }
 
                     messages.append(message_to_append)
                     perf_timing.measure(f"tool completed {tool_call_result.tool_name}")
 
-                    result_dict = {
-                        "tool_call_id": tool_call_result.tool_call_id,
-                        "role": "tool",
-                        "name": tool_call_result.tool_name,
-                        "result": tool_response,
-                    }
+                    # TODO: remove this after FE is ready
+                    # TODO: fix this on the FE side, so we have either `content` or `result` for both streaming and non-streaming responses
+                    if HOLMES_STRUCTURED_OUTPUT_CONVERSION_FEATURE_FLAG:
+                        result_dict = {
+                            "tool_call_id": tool_call_result.tool_call_id,
+                            "role": "tool",
+                            "name": tool_call_result.tool_name,
+                            "content": tool_response_content,
+                        }
+                    else:
+                        result_dict = {
+                            "tool_call_id": tool_call_result.tool_call_id,
+                            "role": "tool",
+                            "name": tool_call_result.tool_name,
+                            "result": tool_call_result.result.model_dump(),
```

Comment:
Shouldn't this be a string with the same value of `_get_formatted_tool_call_response_content`